src/ua/main.cpp

```diff
@@ -529,6 +538,63 @@ void printEnvironmentInfo() {
   }
 }
 
+// There is currently the possibility of a race condition if a chunk
+// upload timed-out.  It's possible that a second upload succeeds,
+// has the chunk marked as "complete" and then the first request makes
+// its way through the queue and marks the chunk as pending again.
+// Since we are just about to close the file, we'll check to see if any
+// chunks are marked as pending, and if so, we'll retry them.
+void check_for_complete_chunks(vector<File> &files) {
+  for (int currCheckNum=0; currCheckNum < NUM_CHUNK_CHECKS; ++currCheckNum){
+    map<string, JSON> fileDescriptions;
+    while (!chunksFinished.empty()) {
+      Chunk *c = chunksFinished.consume();
+
+      // Cache file descriptions so we only have to do once per file,
+      // not once per chunk.
+      if (fileDescriptions.find(c->fileID) == fileDescriptions.end())
+          fileDescriptions[c->fileID] = fileDescribe(c->fileID);
+
+      if (!is_chunk_complete(c, fileDescriptions[c->fileID])) {
+          chunksToUpload.produce(c);
+      }
+    }
+    // All of the chunks were marked as complete, so let's exit and we
+    // should be safeish to close the file.
+    if(chunksToUpload.size() == 0)
+      return;
+
+    // Upload the chunks which weren't marked as complete.
+    DXLOG(logINFO) << " upload...";
+    for (int i = 0; i < opt.uploadThreads; ++i) {
+      uploadThreads.push_back(boost::thread(uploadChunks, boost::ref(files)));
+    }
+
+    // Wait for the upload threads to complete.
+    DXLOG(logINFO) << " upload...";
+    for (int i = 0; i < (int) uploadThreads.size(); ++i) {
+      uploadThreads[i].join();
```

Comment:
I think we have to interrupt the threads, or else they never exit.

Looks like the current code waits for the monitor thread to finish before doing interrupt & join worker threads. Monitor exits when finished() is true.